### PR TITLE
Xen: reading and writing registers

### PR DESCRIPTION
**Pull request by arnabcs17b006**

### CODE REVIEW

File: src/driver/xen.rs [hunk 1]
```diff
@@ -112,58 +114,133 @@ impl Introspectable for Xen {
...
+    fn write_registers(&self, vcpu: u16, reg: Registers) -> Result<(), Box<dyn Error>> {
+        let (mut buffer, size) = self.xc.domain_hvm_getcontext(self.domid)?;
+        //println!("{}", size);
+        let mut offset: u32 = 0;
+        let hvm_save_cpu =
+            unsafe { mem::MaybeUninit::<__HVM_SAVE_TYPE_CPU>::zeroed().assume_init() };
+        let hvm_save_code_cpu: u16 = mem::size_of_val(&hvm_save_cpu.c).try_into().unwrap();
+        //let hvm_save_type_cpu = type_of(&hvm_save_cpu.t).try_into().unwrap();
+        let mut cpu: *mut hvm_hw_cpu = std::ptr::null_mut();
+        unsafe {
```
Comment (rageagainsthepc) [inline src/driver/xen.rs:171]: You should never use unsafe blocks in libmicrovmi. The whole point of having crates like `xenctrl` or `kvmi` is to provide safe bindings. Anything that is unsafe should be done in there.

**Comment by Wenzel** [comment]
@arnabcs17b006 please rebase

File: examples/regs-dump.rs [hunk 2]
```diff
@@ -19,7 +19,8 @@ fn main() {
 
     println!("pausing the VM");
     drv.pause().expect("Failed to pause VM");
-    let total_vcpu_count: u16 = drv.get_vcpu_count().expect("Failed to get vcpu count");
+    //let total_vcpu_count: u16 = drv.get_vcpu_count().expect("Failed to get vcpu count");
+    let total_vcpu_count: u16 = 1;
```
Comment (Wenzel) [inline examples/regs-dump.rs:23]: why is this value hardcoded to 1 ?

File: src/driver/kvm.rs [hunk 3]
```diff
@@ -137,16 +137,13 @@ impl<T: KVMIntrospectable> Introspectable for Kvm<T> {
...
-            efer: sregs.efer,
-            apic_base: sregs.apic_base,
             cs: sregs.cs.into(),
             ds: sregs.ds.into(),
             es: sregs.es.into(),
             fs: sregs.fs.into(),
             gs: sregs.gs.into(),
             ss: sregs.ss.into(),
             tr: sregs.tr.into(),
-            ldt: sregs.ldt.into(),
```
Comment (Wenzel) [inline src/driver/kvm.rs:149]: why are you removing these values from the KVM driver ???

File: src/driver/xen.rs [hunk 4]
```diff
@@ -112,58 +113,111 @@ impl Introspectable for Xen {
...
             },
             tr: SegmentReg {
-                base: 0,
-                limit: 0,
-                selector: 0,
-            },
-            ldt: SegmentReg {
-                base: 0,
-                limit: 0,
-                selector: 0,
```
Comment (Wenzel) [inline src/driver/xen.rs:162]: What happened to ldt ?

**Review by Wenzel — CHANGES_REQUESTED** [review]

**Comment by Wenzel** [comment]
@arnabcs17b006 also please rebase

File: examples/regs-dump.rs [hunk 5]
```diff
@@ -33,7 +33,8 @@ fn main() {
 
     println!("pausing the VM");
     drv.pause().expect("Failed to pause VM");
-    let total_vcpu_count: u16 = drv.get_vcpu_count().expect("Failed to get vcpu count");
+    //let total_vcpu_count: u16 = drv.get_vcpu_count().expect("Failed to get vcpu count");
```
Comment (Wenzel) [inline examples/regs-dump.rs:36]: which is this code commented ?
you replace an expect by an unwrap.

**Review by Wenzel — CHANGES_REQUESTED** [review]

**Pull request closed without merging**